### PR TITLE
740-2 account package

### PR DESCRIPTION
**Pull request by Thykof**
- [ ] handle version missing in the yaml

### CODE REVIEW

File: pkg/wallet/account/account.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		Nickname:     nickname,
+		Address:      *address,
+		Salt:         salt,
+		Nonce:        nonce,
+		CipheredData: cipheredData,
+		PublicKey:    publicKey,
+	}, nil
+}
+
+func NewFromPrivateKey(password *memguard.LockedBuffer, nickname string, privateKeyText *memguard.LockedBuffer) (*Account, error) {
```
Comment (Thykof) [inline pkg/wallet/account/account.go:89]: there is probably a way to avoid code duplication between those to constructor functions

File: pkg/app/app.go [hunk 2]
```diff
@@ -90,50 +88,8 @@ func (a *WalletApp) Hide() {
...
-	}
-
-	if filePath == "" {
-		return selectFileResult{Err: utils.ErrNoFile, CodeMessage: utils.ErrNoFile}
-	}
-
-	wallet, loadErr := wallet.LoadFile(filePath)
-	if loadErr != nil {
-		fmt.Printf("error while loading file: %v", loadErr.Err)
-		return selectFileResult{Err: loadErr.Err.Error(), CodeMessage: loadErr.CodeErr}
```
Comment (peterjah) [inline pkg/app/app.go:117]: why deleting all theses methods ?? they are binded function that are called by the frontend

Comment (Thykof) [reply pkg/app/app.go]: you're right

File: pkg/prompt/import.go [hunk 3]
```diff
@@ -38,16 +39,16 @@ func handleImportFile(prompterApp WalletPrompterInterface, filePath string) (*wa
 	}
 
 	// Validate nickname
-	if !wallet.NicknameIsValid(account.Nickname) {
+	if !account.NicknameIsValid(acc.Nickname) {
```
Comment (gregLibert) [inline pkg/prompt/import.go:42]: What about providing a wallet.CheckAccount(acc) function that will perform all checks ?

File: pkg/types/encrypted_private_key.go [hunk 4]
```diff
@@ -78,11 +79,11 @@ func (a *EncryptedPrivateKey) UnmarshalBinary(data []byte) error {
 }
 
 // Sign signs the given data using the private key.
-// password is destroyed just after being used even if an error occurs.
-func (e *EncryptedPrivateKey) Sign(password *memguard.LockedBuffer, salt, nonce, data []byte) ([]byte, error) {
+// Guarded password is destroyed.
+func (e *EncryptedPrivateKey) Sign(guardedPassword *memguard.LockedBuffer, salt, nonce, data []byte) ([]byte, error) {
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:83]: Why did you rename *password* to *guardedPassword*? The code readability has been made heavier without the information provided being important for the reader, because anyway you added an implementation comment explaining that the password will be destroyed.

File: pkg/types/encrypted_private_key.go [hunk 5]
```diff
@@ -92,9 +93,9 @@ func (e *EncryptedPrivateKey) Sign(password *memguard.LockedBuffer, salt, nonce,
 	return append([]byte{EncryptedPrivateKeyLastVersion}, ed25519.Sign(privateKeyInClear.Bytes(), digest[:])...), nil
 }
 
-// PublicKey returns the public key corresponding to the private key.
-func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, nonce []byte) (*PublicKey, error) {
-	privateKeyInClear, err := PrivateKey(password, salt, nonce, e.Data)
+// PublicKey returns the public key corresponding to the private key. Guarded password is destroyed.
+func (e *EncryptedPrivateKey) PublicKey(guardedPassword *memguard.LockedBuffer, salt, nonce []byte) (*PublicKey, error) {
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:97]: The same goes here.

File: pkg/types/encrypted_private_key.go [hunk 6]
```diff
@@ -112,9 +113,40 @@ func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, n
 	}, nil
 }
 
-func PrivateKey(password *memguard.LockedBuffer, salt, nonce, encryptedKey []byte) (*memguard.LockedBuffer, error) {
-	aeadCipher, secretKey, err := crypto.NewSecretCipher(password.Bytes(), salt[:])
-	defer password.Destroy()
+// PrivateKeyTextInClear returns the private key in clear. Guarded password is destroyed.
+func (e *EncryptedPrivateKey) PrivateKeyTextInClear(guardedPassword *memguard.LockedBuffer, salt, nonce, encryptedKey []byte) (*memguard.LockedBuffer, error) {
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:117]: It's the same here.

File: pkg/types/encrypted_private_key.go [hunk 7]
```diff
@@ -112,9 +113,40 @@ func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, n
...
+	privateKeyInClear, err := PrivateKey(guardedPassword, salt, nonce, encryptedKey)
+	if err != nil {
+		return nil, err
+	}
+
+	seed := ed25519.PrivateKey(privateKeyInClear.Bytes()).Seed()
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:123]: privateKeyInClear.Bytes() has the same value as seed.Bytes(), why do you need to duplicate it?

File: pkg/types/encrypted_private_key.go (hunk 6 above)
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:117]: Why didn't you create a privateKey object with marshal and unmarhsal function as the other objects ?

File: pkg/wallet/account/account.go [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+	"github.com/massalabs/station-massa-wallet/pkg/types/object"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	AccountLastVersion = 1
+)
+
+type Account struct {
+	Version      *uint8                    `yaml:"Version"`
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:21]: Why is it a pointer ?

File: pkg/wallet/account/account.go [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+func New(version *uint8, nickname string, address types.Address, salt [16]byte, nonce [12]byte, cipheredData types.EncryptedPrivateKey, publicKey types.PublicKey) (*Account, error) {
+	if version == nil {
+		version = new(uint8)
+		*version = AccountLastVersion
+	}
+
+	if *version > AccountLastVersion {
+		return nil, fmt.Errorf("unsupported account version: %d", *version)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:37]: you should wrap an invalidArgument sentinel error.

File: pkg/wallet/account/account.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+		return nil, fmt.Errorf("validating address: %w", err)
+	}
+
+	if err := cipheredData.Validate(cipheredData.Version, object.EncryptedPrivateKey); err != nil {
+		return nil, fmt.Errorf("validating encrypted private key: %w", err)
+	}
+
+	if err := publicKey.Validate(publicKey.Version, object.PublicKey); err != nil {
+		fmt.Println("52", publicKey.Version)
+		fmt.Println("53", publicKey.Kind)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:54]: Debug code to be removed ?

File: pkg/wallet/account/account.go [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+		Nickname:     nickname,
+		Address:      address,
+		Salt:         salt,
+		Nonce:        nonce,
+		CipheredData: cipheredData,
+		PublicKey:    publicKey,
+	}, nil
+}
+
+func NewEmpty() Account {
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:69]: Why do you need this function?

File: pkg/wallet/account/account.go [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+				Kind:    object.PublicKey,
+				Version: 0x00,
+				Data:    nil,
+			},
+		},
+	}
+}
+
+// NewGenerated generates a new account with a random private key. It destroys the guarded password.
+func NewGenerated(guardedPassword *memguard.LockedBuffer, nickname string) (*Account, error) {
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:100]: Should be named Generate

File: pkg/wallet/account/account.go [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+	version := uint8(AccountLastVersion)
+
+	publicKeyBytes, privateKeyBytes, err := ed25519.GenerateKey(nil)
+	if err != nil {
+		return nil, fmt.Errorf("generating ed25519 key-pair: %w", err)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:105]: You should also wrap a sentinel error.

File: pkg/wallet/account/account.go [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+			},
+		},
+	}
+}
+
+// NewGenerated generates a new account with a random private key. It destroys the guarded password.
+func NewGenerated(guardedPassword *memguard.LockedBuffer, nickname string) (*Account, error) {
+	version := uint8(AccountLastVersion)
+
+	publicKeyBytes, privateKeyBytes, err := ed25519.GenerateKey(nil)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:103]: You should add a security comment explaining that the entropy is read from [crypto/rand.Reader](https://pkg.go.dev/crypto/rand#Reader) or even better, diretcly use rand as you already import it for the salt and the nonce

File: pkg/wallet/account/account.go [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+	}
+
+	aeadCipher, secretKey, err := crypto.NewSecretCipher(guardedPassword.Bytes(), salt[:])
+	guardedPassword.Destroy()
+	if err != nil {
+		return nil, fmt.Errorf("creating secret cipher: %w", err)
+	}
+
+	// secretBuffer is the secret key in clear without the version.
+	secretBuffer := memguard.NewBufferFromBytes(privateKeyBytes.Seed())
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:127]: You're mixing secret key and provate key. here you should talk about the private keyu instead.

File: pkg/wallet/account/account.go [hunk 16]
```diff
@@ -0,0 +1,265 @@
...
+	Version      *uint8                    `yaml:"Version"`
+	Nickname     string                    `yaml:"Nickname,omitempty"`
+	Address      types.Address             `yaml:"Address,omitempty"`
+	Salt         [16]byte                  `yaml:"Salt,flow"`
+	Nonce        [12]byte                  `yaml:"Nonce,flow"`
+	CipheredData types.EncryptedPrivateKey `yaml:"CipheredData,flow"`
+	PublicKey    types.PublicKey           `yaml:"PublicKey,flow"`
+}
+
+func New(version *uint8, nickname string, address types.Address, salt [16]byte, nonce [12]byte, cipheredData types.EncryptedPrivateKey, publicKey types.PublicKey) (*Account, error) {
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:30]: Why address, types.EncryptedPrivateKey, publicKey are not pass by refenrece instead of duplicating them ?

Furthermore cipheredData is obscure, you should use encryptedPrivateKey as you are using address or publickey.

File: pkg/wallet/account/account.go [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+	if err != nil {
+		seedBuffer.Destroy()
+		return nil, fmt.Errorf("decoding base58 private key: %w", err)
+	}
+
+	privateKey := memguard.NewBufferFromBytes(ed25519.NewKeyFromSeed(seedBuffer.Bytes()))
+	seedBuffer.Destroy()
+
+	encryptedSecret := crypto.SealSecret(aeadCipher, nonce[:], privateKey)
+	secretKey.Destroy()
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:184]: I have the feeling that the same code (sealing a private key) is called multiple times. Can you create a hekper to do that only at one place.

File: pkg/wallet/account/account.go [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+		return fmt.Errorf("missing ciphered data")
+	}
+
+	if len(a.PublicKey.Data) == 0 {
+		return fmt.Errorf("missing public key")
+	}
+
+	// if a.Version == nil {
+	// 	return fmt.Errorf("missing version")
+	// }
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:258]: dead code to be removed ?

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: pkg/wallet/account/account.go (hunk 8 above)
Comment (Thykof) [reply pkg/wallet/account/account.go:24]: idk. I need help on the version

File: pkg/prompt/import.go (hunk 3 above)
Comment (Thykof) [reply pkg/prompt/import.go:42]: good idea, I'll try to do that in the following PR on wallet package

File: pkg/wallet/account/account.go (hunk 9 above)
Comment (Thykof) [reply pkg/wallet/account/account.go:40]: what about `object.ErrUnsupportedVersion`?

File: pkg/wallet/account/account.go (hunk 11 above)
Comment (Thykof) [reply pkg/wallet/account/account.go:72]: to do this:

```go
acc := account.NewEmpty()
err = acc.Unmarshal(data)
if err != nil {
  return nil, fmt.Errorf("%w: %w", ErrUnmarshalAccount, err)
}
```

File: pkg/wallet/account/account.go (hunk 8 above)
Comment (Thykof) [reply pkg/wallet/account/account.go:27]: I need it to be a pointer because it's a mandatory field in the yaml, and the yaml unmarshal will put value nil and so that I can throw an error if the value is nil after unmarshaling the yaml

File: pkg/wallet/account/account.go (hunk 11 above)
Comment (gregLibert) [reply pkg/wallet/account/account.go:81]: IMHO this need is the result of the Struct  Account which is haphazard: optional objects and consequential objects (to reduce memory footprint and duplication) should be pointers, whereas objects with default values should be allocations.

File: pkg/wallet/account/account.go (hunk 9 above)
Comment (gregLibert) [reply pkg/wallet/account/account.go]: I was more thinking of:
```golang
var ErrInvalidParameter = errors.New("invalid parameter")
...
if version > AccountLastVersion || version == UnknownVersion {
   return nil, return fmt.Errorf("%w: invalid version", ErrInvalidParameter)
}
```
This way, on the caller side you can easily detect that an error is about an invalid parameter, to return an http error code 422 for instance, by doing:
```golang
err := account.New(...)
if errors.Is(err, ErrInvalidParameter) {
    // handle the invalid parameter error
}
```

File: pkg/wallet/account/account.go (hunk 8 above)
Comment (gregLibert) [reply pkg/wallet/account/account.go:27]: Your reason seems shaky: using uint8(0) as UnknownVersion value would avoid using a pointer and will be consistent with the default value of an uint8.

File: pkg/types/encrypted_private_key.go (hunk 7 above)
Comment (gregLibert) [reply pkg/types/encrypted_private_key.go:124]: It's weird, looking at the code of the Seed function, it should return the private key bytes: https://cs.opensource.google/go/go/+/refs/tags/go1.21.3:src/crypto/ed25519/ed25519.go;l=75

File: pkg/wallet/account/account.go (hunk 8 above)
Comment (Thykof) [reply pkg/wallet/account/account.go:27]: yes but value 0 for a version is a correct value, our standard wallet file format first version is 0.

Comment (Thykof) [reply pkg/wallet/account/account.go:27]: let's take 0 as unknown version

`UnknownVersion := uint8(0)`

File: pkg/wallet/account/account.go (hunk 11 above)
Comment (Thykof) [reply pkg/wallet/account/account.go:81]: no need this function

acc := Account{}

**Review by peterjah — APPROVED** [review]

File: pkg/wallet/account/account.go [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+	"github.com/btcsuite/btcutil/base58"
+	"github.com/massalabs/station-massa-wallet/pkg/crypto"
+	"github.com/massalabs/station-massa-wallet/pkg/types"
+	"github.com/massalabs/station-massa-wallet/pkg/types/object"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	AccountLastVersion    = 1
+	AccountUnknownVersion = 0
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:19]: Account prefix is useless as you're already in an Account package.

File: pkg/types/encrypted_private_key.go [hunk 20]
```diff
@@ -29,74 +31,81 @@ func (a *EncryptedPrivateKey) validate() error {
...
-// password is destroyed just after being used even if an error occurs.
+// Sign signs the given data using the private key. Password is destroyed.
 func (e *EncryptedPrivateKey) Sign(password *memguard.LockedBuffer, salt, nonce, data []byte) ([]byte, error) {
 	digest := blake3.Sum256(data)
 
-	privateKeyInClear, err := PrivateKey(password, salt, nonce, e.Data)
+	privateKeyInClear, err := privateKey(password, salt, nonce, e.Data)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("Sign: %w", err)
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:96]: Error message should start with capital letters in go. Furthermore your error is not giving context, it simply contains the name of the function.

my suggestion:
```suggestion
		return nil, fmt.Errorf("failed to get private key: %w", err)
```

File: pkg/types/encrypted_private_key.go [hunk 21]
```diff
@@ -29,74 +31,81 @@ func (a *EncryptedPrivateKey) validate() error {
...
 }
 
-// PublicKey returns the public key corresponding to the private key.
+// PublicKey returns the public key corresponding to the private key. Password is destroyed.
 func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, nonce []byte) (*PublicKey, error) {
-	privateKeyInClear, err := PrivateKey(password, salt, nonce, e.Data)
+	privateKeyInClear, err := privateKey(password, salt, nonce, e.Data)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("PublicKey: %w", err)
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:108]: This is also true here.

File: pkg/types/encrypted_private_key.go [hunk 22]
```diff
@@ -112,7 +121,37 @@ func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, n
 	}, nil
 }
 
-func PrivateKey(password *memguard.LockedBuffer, salt, nonce, encryptedKey []byte) (*memguard.LockedBuffer, error) {
+// PrivateKeyTextInClear returns the private key in clear. Password is destroyed.
+func (e *EncryptedPrivateKey) PrivateKeyTextInClear(password *memguard.LockedBuffer, salt, nonce, encryptedKey []byte) (*memguard.LockedBuffer, error) {
+	privateKeyInClear, err := privateKey(password, salt, nonce, encryptedKey)
+	if err != nil {
+		return nil, fmt.Errorf("PrivateKeyTextInClear: %w", err)
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:128]: Same here.

File: pkg/types/encrypted_private_key.go [hunk 23]
```diff
@@ -112,7 +121,37 @@ func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, n
...
+	}
+
+	seed := ed25519.PrivateKey(privateKeyInClear.Bytes()).Seed()
+	privateKeyInClear.Destroy()
+
+	seedBuffer := memguard.NewBufferFromBytes(seed)
+
+	privateKey := e.Kind.Prefix() + base58.CheckEncode(seedBuffer.Bytes(), e.Version)
+	privateKeyBuffer := memguard.NewBufferFromBytes([]byte(privateKey))
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:137]: Why seedBuffer is not destroyed?

File: pkg/types/encrypted_private_key.go [hunk 24]
```diff
@@ -112,7 +121,37 @@ func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, n
...
+	return privateKeyBuffer, nil
+}
+
+// PasswordIsValid returns true if the password is valid for the account. It destroys the password.
+func (e *EncryptedPrivateKey) PasswordIsValid(password *memguard.LockedBuffer, salt, nonce, encryptedKey []byte) bool {
+	privateKeyInClear, err := privateKey(password, salt, nonce, e.Data)
+	if err != nil {
+		return false
+	}
+	privateKeyInClear.Destroy()
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:148]: This line seems to be wronly formatted. Can you check the linter?

File: pkg/types/encrypted_private_key.go [hunk 25]
```diff
@@ -112,7 +121,37 @@ func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, n
...
+
+	return err == nil
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:150]: err is necessarly nil by construction here.

File: pkg/types/encrypted_private_key.go [hunk 26]
```diff
@@ -112,7 +121,37 @@ func (e *EncryptedPrivateKey) PublicKey(password *memguard.LockedBuffer, salt, n
...
+
+	return privateKeyBuffer, nil
+}
+
+// PasswordIsValid returns true if the password is valid for the account. It destroys the password.
+func (e *EncryptedPrivateKey) PasswordIsValid(password *memguard.LockedBuffer, salt, nonce, encryptedKey []byte) bool {
+	privateKeyInClear, err := privateKey(password, salt, nonce, e.Data)
+	if err != nil {
+		return false
+	}
```
Comment (gregLibert) [inline pkg/types/encrypted_private_key.go:147]: The logic of this function is weird. If the salt or the nonce is wrong, but the password is ok, you will also get an error. and the function PasswordIsValid will return false...
Furthermore, are you sure that except from an invalid salt, nonce or password, no other error are returned ?

File: pkg/wallet/account/account.go [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+)
+
+var (
+	ErrInvalidPrivateKey = errors.New("invalid private key")
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:23]: Why the private key is having it's own error? What about reusing the InvalidParameter aggregation instead?

File: pkg/wallet/account/account.go [hunk 28]
```diff
@@ -0,0 +1,246 @@
...
+	ErrGeneratingKeyPair = errors.New("generating ed25519 key-pair")
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:24]: This error is too specific and will make your code more complex to read in the error handling section.
I suggest to do the following:
```golang
var ErrRuntime = errors.New("runtime execution")
...
publicKeyBytes, privateKeyBytes, err := ed25519.GenerateKey(rand.Reader)
if err != nil {
	return nil, fmt.Errorf("key generation failed: %w: %w", ErrRuntime, err)
}
...
```
That way you'll be able to catch the ErrRuntime to return http code 500.

Note: the second %w is optional, you could use %v instead, but doing it you wrap the underlying error, exposing it to the caller if at somepoint he will need to differenciate different errors.

File: pkg/wallet/account/account.go [hunk 29]
```diff
@@ -0,0 +1,246 @@
...
+	ErrInvalidParameter  = errors.New("invalid parameter")
+)
+
+type Account struct {
+	Version      uint8                      `yaml:"Version"`
+	Nickname     string                     `yaml:"Nickname,omitempty"`
+	Address      *types.Address             `yaml:"Address,omitempty"`
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:31]: Why do youo have omitempty here and not elsewhere ?

File: pkg/wallet/account/account.go [hunk 30]
```diff
@@ -0,0 +1,246 @@
...
+	version uint8,
+	nickname string,
+	address *types.Address,
+	salt [16]byte,
+	nonce [12]byte,
+	encryptedPrivateKey *types.EncryptedPrivateKey,
+	publicKey *types.PublicKey,
+) (*Account, error) {
+	if version > AccountLastVersion {
+		return nil, fmt.Errorf("%w: %d", ErrInvalidParameter, version)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:48]: The error message will be "invalid parameter: 0". Can you change it to something easier to understand ?

File: pkg/wallet/account/account.go [hunk 31]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	if !NicknameIsValid(nickname) {
+		return nil, fmt.Errorf("%w: %s", ErrInvalidNickname, nickname)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:52]: This is also true here and nickname is a parameter

File: pkg/wallet/account/account.go [hunk 32]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	if err := address.Validate(address.Version, object.UserAddress, object.SmartContractAddress); err != nil {
+		return nil, fmt.Errorf("validating address: %w", err)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:56]: You should use invalid parameter sentinel error

File: pkg/wallet/account/account.go [hunk 33]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	if err := encryptedPrivateKey.Validate(encryptedPrivateKey.Version, object.EncryptedPrivateKey); err != nil {
+		return nil, fmt.Errorf("validating encrypted private key: %w", err)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:60]: Ok, I guess you can review all errors

File: pkg/wallet/account/account.go [hunk 34]
```diff
@@ -0,0 +1,246 @@
...
+	return a.CipheredData.PrivateKeyTextInClear(password, a.Salt[:], a.Nonce[:], a.CipheredData.Data)
+}
+
+// Sign signs the data with the private key and destroys the password.
+func (a *Account) Sign(password *memguard.LockedBuffer, data []byte) ([]byte, error) {
+	return a.CipheredData.Sign(password, a.Salt[:], a.Nonce[:], data)
+}
+
+func (a *Account) Marshal() ([]byte, error) {
+	yamlMarshaled, err := yaml.Marshal(a)
```
Comment (gregLibert) [inline pkg/wallet/account/account.go:206]: why don't you directly return yaml.Marshal(a) ? Except if you want to wrap it using  sentinel error, but you are not doing so...

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: pkg/wallet/account/account.go (hunk 29 above)
Comment (Thykof) [reply pkg/wallet/account/account.go:31]: nickname and address are optional fields

**Review by gregLibert — APPROVED** [review]
